Add compose (#19)
asreview/asreview-datatools: README.md

[![PyPI version](https://badge.fury.io/py/asreview-datatools.svg)](https://badge.fury.io/py/asreview-datatools) [![Downloads](https://pepy.tech/badge/asreview-datatools)](https://pepy.tech/project/asreview-datatools) ![PyPI - License](https://img.shields.io/pypi/l/asreview-datatools) ![Deploy and release](https://github.com/asreview/asreview-datatools/workflows/Deploy%20and%20release/badge.svg) ![Build status](https://github.com/asreview/asreview-datatools/workflows/test-suite/badge.svg) [![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.6625879.svg)](https://doi.org/10.5281/zenodo.6625879)

ASReview Datatools is an extension to [ASReview

> LAB](https://github.com/asreview/asreview) that can be used for:

Reviewer: ```suggestion
LAB](https://github.com/asreview/asreview) that can be used to:
```

> - [**Describing**](#describe) basic properties of a dataset (e.g., number of papers, number of inclusions,
> the amount of missing data and duplicates)
> - [**Converting**](#convert) file formats via the command line
> - [**Deduplication**](#dedup): cleaning your (input) data by removing duplicate records.
> - [**Composing**](#compose) datasets with different (or no) labels into a single dataset

Reviewer: ```suggestion
- [**Describe**](#describe) basic properties of a dataset (e.g., number of papers, number of inclusions,
the amount of missing data and duplicates)
- [**Convert**](#convert) file formats via the command line
- [**Deduplicate**](#dedup) data based on properties of the data
- [**Compose**](#compose) a single (labeled, partly labeled, or unlabeled) dataset from multiple datasets
```

[ASReview-statistics](https://pypi.org/project/asreview-statistics/). ASReview
datatools is available for version 1 or later. Use ASReview statistics for
versions 0.x.

LAB](https://github.com/asreview/asreview) that can be used for:
- [**Describing**](#describe) basic properties of a dataset (e.g., number of papers, number of inclusions,
the amount of missing data and duplicates)
- [**Converting**](#convert) file formats via the command line
- [**Deduplication**](#dedup): cleaning your (input) data by removing duplicate records.
- [**Composing**](#compose) datasets with different (or no) labels into a single dataset

ASReview
datatools is available for ASReview Lab **v1.1+** and requires Python 3.7+.
If you are using ASReview Lab v0.x, use [ASReview-statistics](https://pypi.org/project/asreview-statistics/) instead of ASReview datatools.
---
## Installation

The ASReview Datatools extensions requires Python 3.7+ and [ASReview
LAB](https://github.com/asreview/asreview) version 1 or later.

> 

Reviewer: This should remain here. These are installation instructions. Please also check wording.

asreview data NAME_OF_TOOL
```

where `NAME_OF_TOOL` is the name of one of the tools below (`describe`, `convert`, `compose` or `dedup`)
followed by positional arguments and optional arguments.

Each tool has its own help description which is available with

```bash
asreview data NAME_OF_TOOL -h
```

> ---
> ## Describe

Reviewer: ```suggestion
## Commands

### Describe
```

Reviewer: Or tools

asreview data describe benchmark:van_de_schoot_2017 -o output.json
```

<details>
  <summary>Click to see output:</summary>

> 

Reviewer: Please stick to markdown

  }
}
```
</details>

> ## Convert

Reviewer: Mwhaaa, any arguments?

### Data Compose (Experimental)
Compose is where datasets with different labels (or no labels) can be assembled into a single dataset.

❗ Compose is an experimental feature. We would love to hear your feedback. Please keep in mind that this feature can change in the future. 

#### Data format
Your data files need to be in tabular file format or RIS file format.
The output file and all input files should be in the same format.

- **Tabular file format:**
Supported tabular file formats are `.csv`, `.tab`, `.tsv` or `.xlsx`.
Make sure your column names adhere to the predetermined set of [accepted column names](https://asreview.readthedocs.io/en/latest/data_format.html).

- **RIS file format:**
RIS file formats have `.ris` or `.txt` as extension and are used by digital libraries.
Read [how to format](https://asreview.readthedocs.io/en/latest/data_format.html) your RIS files.
ASReview converts the labeling decisions in RIS files to a binary variable: irrelevant as `0` and relevant as `1`.

Records marked as unseen or with missing labeling decisions, are converted to `-1` by ASReview.

#### Run script
Assume you have `MY_DATASET_1.ris` from which you want to keep all existing labels

> and `MY_DATASET_2.ris` which you want to mark as unlabeled. Compose into a single dataset:

Author: @Rensvandeschoot I chose "mark" on purpose because you can also choose to have them unlabeled even though these records are already labeled in the input dataset `MY_DATASET_2.ris`.

asreview data dedup benchmark:van_de_schoot_2017 -o van_de_schoot_2017_dedup.csv
```

> ## Compose
> Compose is where datasets with different labels (or no labels) can be assembled into a single dataset.

Reviewer: ```suggestion
## Compose (Experimental)
Compose is where datasets with different labels (or no labels) can be assembled into a single dataset.

❗ Compose is an experimental feature. We would love to hear your feedback. Please keep in mind that this feature can change in the future. 
```

### Data format
Your data files need to be in tabular file format or RIS file format.

> All input files should use the same format.

Reviewer: ```suggestion
All input files should be in the same format.
```

- **Tabular file format:**
Supported tabular file formats are `.csv`, `.tab`, `.tsv` or `.xlsx`.
Make sure your column names adhere to the predetermined set of [accepted column names](https://asreview.readthedocs.io/en/latest/data_format.html).

- **RIS file format:**
RIS file formats have `.ris` or `.txt` as extension and are used by digital libraries.
Read [how to format](https://asreview.readthedocs.io/en/latest/data_format.html) your RIS files.
ASReview converts the labeling decisions in RIS files to a binary variable: irrelevant as `0` and relevant as `1`.

Records marked as unseen or with missing labeling decisions, are converted to `-1` by ASReview.

### Run script

> Navigate to the directory where your data files are located:
> ```bash
> cd Parent_directory
> ```

Reviewer: ```suggestion
```

Assume you have `MY_DATASET_1.ris` from which you want to keep all existing labels
and `MY_DATASET_2.ris` which you want to mark as unlabeled. Compose into a single dataset:
```bash
asreview data compose composed_output.ris -l MY_DATASET_1.ris -u MY_DATASET_2.ris
```
The resulting dataset is exported to `composed_output.ris`.

The output path (`composed_output.ris` in the example) should always be specified.
It is followed by optional arguments for:
- Input files
- Persistent identifier (PID) used for deduplication
- Resolving conflicting labels

#### Input files
Overview of possible input files and corresponding properties, use at least one of the following arguments:

| Arguments            | Action                                     |
|----------------------|--------------------------------------------|
| `--relevant`, `-r`   | Label all records in dataset `relevant`.   |
| `--irrelevant`, `-i` | Label all records in dataset `irrelevant`. |
| `--labeled`, `-l`    | Keep existing labels in dataset.           |

> | `--unlabeled`, `-u`  | Remove all labels in dataset.              |

Reviewer: Nice

#### Persistent identifier

> Duplicate checking is based on title/abstract and a persistent identifier (PID), see [source code](https://github.com/asreview/asreview/blob/master/asreview/data/base.py#L453).

Reviewer: ```suggestion
Duplicate checking is based on title/abstract and a persistent identifier (PID).
```

|----------------|----------------------------------------------------------------------------------------|
| `continue`     | Continue, using `--priority` to determine which label to keep                          |
| `keep`         | Keep all labels for duplicate records with inconsistent labels (ignoring `--priority`) | 
| `abort`        | Abort                                                                                  |

### Example
```bash
asreview data compose composed_output.ris -l MY_DATASET_1.ris -u MY_DATASET_2.ris -o uir -c abort
```
Above command will compose a dataset from `MY_DATASET_1.ris` and `MY_DATASET_2.ris`.
The labels from `MY_DATASET_1.ris` are kept and all records from `MY_DATASET_2.ris` are marked as unlabeled.
In case any duplicate ambiguously labeled records exist, either within a dataset or across the datasets:
- Unlabeled is prioritized over irrelevant and relevant labels.
- Irrelevant labels are prioritized over relevant labels.

If there are any such conflicting/contradictory labels, the user is warned, the conflicting records are shown and the script is aborted.

### Tutorials

> Several [tutorials](Tutorials.md) are available that show how compose can be used in different scenarios.

Reviewer: Please add this to the new PR

### Note
When the composed dataset is exported to RIS file format you may get a warning similar to:
```bash
C:\Python310\lib\site-packages\rispy\writer.py:114: UserWarning: label `included` not exported
warnings.warn(UserWarning(f"label `{label}` not exported"))
```
Despite the warning, the labels will still be exported, so this warning can be ignored.

> ---

Reviewer: Suppress the warning instead ;) ?
